### DISCUSSION TITLE
Variations of CRC

### DISCUSSION THREAD

**Discussion by Aetherinox** [nullfx/NullFX.CRC#30] [General]
I've tested out all of the algorithms, and I've been comparing them to CyberChef since that's the app I use the most (via https://gchq.github.io/CyberChef/)

Is CyberChef using a different type of CRC algo? Or am I missing a vital setting in yours. Because I've tested out about every variation of CRC16 I can, and I never get the values that I get in CyberChef.

However, CRC32 seems to be fine.

**Comment by Aetherinox** [comment]
Alright, it was a two step issue. One, I had an error in my hex string formatting. Second, your CRC-8 matches with CyberChef's CRC-8/DVB-S2. 

I've known about a few of the variations of CRC, but I went and looked at a total list and had no damn clue there's so many types.

The default for Cyber Chef is CRC 8 tab. Once I plugged those numbers in, I got the same result for CRC 8 in both programs.

Seriously had no clue there were so many. Sorry for the bother.

**Comment by nullfx** [comment]
🤣 it can be confusing for sure, glad you got it sorted, cheers!

here's another good crc website (with admittedly a lot more implementations that I have in the library): https://crccalc.com/

**Comment by Aetherinox** [comment]
Thanks.

That's actually crazy. The website I found yesterday showed that there was like 6 variants of CRC16. The website you sent says there's like 23 damn variants.

Never seen so many variants for a single hash algorithm.

Even CyberChef one gives you like 6 variants for CRC8, whereas with 16. 32 you only get one.

**Reply by nullfx** [reply]
ya there are a lot of ways to do error detection. a lot of smart people have come up with different ways to calculate it. some are better for certain scenarios than others, so that's why there are so many different kinds out there.

here's a pretty good video on the subject (I thought so anyway, could be really boring LOL depending on whether this kind of thing interests you or not): https://www.youtube.com/watch?v=izG7qT0EpBw

**Reply by Aetherinox** [reply]
Thanks, that video is actually cool.

I geek out on this type of stuff. When I bought my first Yubikeys, I spent almost a full month figuring out how it works, the YubiCloud algorithm, wrote my own PHP based API for OTP generation.

Few weeks ago I was doing RSA keys, figuring out the math for how they are generated and reversing them.

Now my newest craze is hashing, checksums, algorithms, seeing how each one works. So I've been messing with CRC, MD5, SHA-3, Blake, and even Argon.

Seems like every time I find a resource that teaches me a lot, I open about 30 more questions that I have to answer.

**Reply by nullfx** [reply]
well, crc is more for error detection in things like messages used in network or serial communication. its quick and efficient for smaller payloads to validate things haven't changed or been corrupted by environmental conditions coming down the line.

as the data you need to validate gets larger the feasibility of crc to (as ben eater's video touched on), at least the ones that produce a smaller sized crc like crc 16 or 32 to miss bits that may have changed become more likely. So its best to use a different (slower but less likely to miss altered bits over much larger payloads) hashing algorithm like sha. When you think about it, trying to create a checksum / hash of a payload several hundred meg to gigs and represent that using a 16 or 32 bit number, you're bound to have data collisions somewhere as that payload gets larger, so its best to use 128bit+ numbers to represent those hashes using something like md5/sha. the larger the number, the better the precision for the hash (sha3 uses a 512 bit number so it can store quite a hash value of extremely large payloads).
